### ISSUE TITLE
Fix memory leaks in BackendMessageEnvelopeDecoder

### ISSUE THREAD

**Issue by mp911de** [r2dbc/r2dbc-postgresql#157]
```
22:33:19.984 reactor-tcp-nio-10 io.netty.util.ResourceLeakDetector                      LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:349)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:115)
	io.r2dbc.postgresql.message.backend.BackendMessageUtils.readComposite(BackendMessageUtils.java:60)
	io.r2dbc.postgresql.message.backend.BackendMessageUtils.getEnvelope(BackendMessageUtils.java:53)
	io.r2dbc.postgresql.message.backend.BackendMessageEnvelopeDecoder.lambda$apply$0(BackendMessageEnvelopeDecoder.java:65)
	reactor.core.publisher.FluxCreate.subscribe(FluxCreate.java:94)
	reactor.core.publisher.Flux.subscribe(Flux.java:8127)
```

**Comment by mp911de** [comment]
That should get fixed with e5de0b61730ab8e34825f294eddef776a7a094e0.

**Comment by mp911de** [comment]
@Squiry I experimented with Netty's `LengthFieldBasedFrameDecoder` (and copying `nioBuffers` in the decoder when retaining these). It turned out that `LengthFieldBasedFrameDecoder` can replace `BackendMessageEnvelopeDecoder` and solve the memory leaks that showed up.

**Comment by Squiry** [comment]
Netty [documentation](https://netty.io/wiki/reference-counted-objects.html#best-practices-to-avoid-leaks) recommends to run unit/integration tests with PARANOID level of leak detection, have you tried this? Cause I don't think there's a leak in BackendMessageEnvelopeDecoder.
But switch to LengthFieldBasedFrameDecoder is a good idea anyway, I just didn't want to move to netty level since we were going all-reactor here. I'de like to change BackendMessageDecoder to be a netty's ByteToMessageCodec too.

**Comment by Squiry** [comment]
I've run jmh branch with paranoid leak detection and leaks are still there.
Some of them are fixed [here](https://github.com/Squiry/r2dbc-postgresql/tree/byte-buf-leaks), but there are still some leaks when running `PostgresqlTestKit`, I'll investigate this later.

**Comment by mp911de** [comment]
Thanks for having a look. I didn't yet get to further work on the driver as I'm currently on the road. Once the `jmh` branch is merged, we can build PR's on top of it. We also need to merge your SSL PR.

**Comment by Squiry** [comment]
So, last two leaks were in validate connection method and in codecs when bind failes.

**Comment by mangan77** [comment]
Not sure if this is a known leak or not, but I thought I'll post it anyway if it could be of interest:

`2019-09-20 16:23:40.368 ERROR 35672 --- [actor-tcp-nio-2] io.netty.util.ResourceLeakDetector       : LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:349)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:173)
	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:107)
	io.r2dbc.postgresql.util.ByteBufUtils.encode(ByteBufUtils.java:57)
	io.r2dbc.postgresql.codec.StringCodec.doEncode(StringCodec.java:69)
	io.r2dbc.postgresql.codec.StringCodec.doEncode(StringCodec.java:36)
	io.r2dbc.postgresql.codec.AbstractCodec.encode(AbstractCodec.java:75)
	io.r2dbc.postgresql.codec.DefaultCodecs.encode(DefaultCodecs.java:106)
	io.r2dbc.postgresql.ExtendedQueryPostgresqlStatement.bind(ExtendedQueryPostgresqlStatement.java:83)
	io.r2dbc.postgresql.ExtendedQueryPostgresqlStatement.bind(ExtendedQueryPostgresqlStatement.java:76)
	io.r2dbc.postgresql.ExtendedQueryPostgresqlStatement.bind(ExtendedQueryPostgresqlStatement.java:36)
	io.r2dbc.client.Update.bind(Update.java:57)
	com.teliacompany.iptv.imageservice.persistence.repository.ImageReferenceRepositoryR2DBC.insert(ImageReferenceRepositoryR2DBC.java:128)
	com.teliacompany.iptv.imageservice.persistence.repository.ImageReferenceRepositoryR2DBC.lambda$save$1(ImageReferenceRepositoryR2DBC.java:34)
	io.r2dbc.client.Handle.inTransaction(Handle.java:154)
	io.r2dbc.client.R2dbc.lambda$inTransaction$0(R2dbc.java:59)
	io.r2dbc.client.R2dbc.lambda$withHandle$2(R2dbc.java:123)
	reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onNext(MonoFlatMapMany.java:153)
	reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1515)
	reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.signal(MonoDelayUntil.java:221)
	reactor.core.publisher.MonoDelayUntil$DelayUntilTrigger.onComplete(MonoDelayUntil.java:300)
	reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onComplete(MonoIgnoreElements.java:81)
	reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:144)
	reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.checkTerminated(FluxWindowPredicate.java:506)
	reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.drainLoop(FluxWindowPredicate.java:454)
	reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.drain(FluxWindowPredicate.java:398)
	reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.onComplete(FluxWindowPredicate.java:290)
	reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onComplete(MonoFlatMapMany.java:248)
	reactor.core.publisher.FluxWindowPredicate$WindowFlux.checkTerminated(FluxWindowPredicate.java:734)
	reactor.core.publisher.FluxWindowPredicate$WindowFlux.drainRegular(FluxWindowPredicate.java:628)
	reactor.core.publisher.FluxWindowPredicate$WindowFlux.drain(FluxWindowPredicate.java:714)
	reactor.core.publisher.FluxWindowPredicate$WindowFlux.onComplete(FluxWindowPredicate.java:780)
	reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.onNext(FluxWindowPredicate.java:241)
	reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:113)
	reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:320)
	reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:320)
	reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:320)
	reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:826)
	reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.innerNext(FluxConcatMap.java:275)
	reactor.core.publisher.FluxConcatMap$ConcatMapInner.onNext(FluxConcatMap.java:849)
	reactor.core.publisher.FluxGenerate$GenerateSubscription.next(FluxGenerate.java:169)
	io.r2dbc.postgresql.message.backend.BackendMessageDecoder.lambda$decode$1(BackendMessageDecoder.java:133)
	reactor.core.publisher.FluxGenerate$GenerateSubscription.slowPath(FluxGenerate.java:262)
	reactor.core.publisher.FluxGenerate$GenerateSubscription.request(FluxGenerate.java:204)
	reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:1879)
	reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:1753)
	reactor.core.publisher.FluxGenerate.subscribe(FluxGenerate.java:83)
	reactor.core.publisher.Flux.subscribe(Flux.java:7921)
	reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:442)
	reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onNext(FluxConcatMap.java:244)
	reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:192)
	reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:206)
	reactor.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:322)
	reactor.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:344)
	reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:91)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1421)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:697)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:632)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:549)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:511)
	io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:834)`

**Comment by Squiry** [comment]
Looks like the one with failed binding, but you can never tell for sure from a stack trace. I believe it is fixed [here](https://github.com/Squiry/r2dbc-postgresql/commit/c0abd734684799ee1d56beb8a200e299eacba40a).

**Comment by mp911de** [comment]
Deferring encoding to the actual subscription makes sense to avoid early buffer allocation. @Squiry do you want to submit a pull request with your changes?

**Comment by Squiry** [comment]
@mp911de 
Sure, #165 here it is

**Comment by mp911de** [comment]
Fixed via #165.
